modules/charon-core/src/main/java/org/wso2/charon3/core/config/ExtensionBuilder.java

```diff
@@ -0,0 +1,495 @@
+     * @param subAttributeName SubAttributeName.
+     * @return Subattribute URI
+     */
+    protected String getSubAttributeURI(ExtensionAttributeSchemaConfig config, String subAttributeName) {
+
+        if (isRootConfig(config)) {
+            return config.getURI() + ":" + subAttributeName;
+        } else {
+            return config.getURI() + "." + subAttributeName;
+        }
+    }
+
+    /**
+     * Returns true if it is the rootconfig. Else returns false.
+     *
+     * @param config ExtensionAttributeSchemaConfig
+     * @return Returns true if it is the rootconfig. Else returns false.
+     */
+    protected boolean isRootConfig(ExtensionAttributeSchemaConfig config) {
+
+        return getURI().equals(config.getURI());
+    }
+
+    /*
+     * Knows how to build a complex attribute
+     *
+     * @param config
+     */
+    protected void buildComplexAttributeSchema(ExtensionAttributeSchemaConfig config,
+                                               Map<String, AttributeSchema> attributeSchemas, Map<String,
+            ExtensionAttributeSchemaConfig> customConfig) throws InternalErrorException {
+
+        if (!attributeSchemas.containsKey(config.getURI())) {
+            String[] subAttributes = config.getSubAttributes();
+            for (String subAttribute : subAttributes) {
+                ExtensionAttributeSchemaConfig subAttribConfig = customConfig.get(getSubAttributeURI(config,
+                        subAttribute));
+                if (subAttribConfig == null) {
+                    String error = String.format("Error adding subattribute %s to attribute %s. Error in SCIM2 " +
+                            "extension schema config format.", subAttribute, config.getURI());
+                    throw new InternalErrorException(error);
+                }
+
+                if (!subAttribConfig.getType().equals(SCIMDefinitions.DataType.COMPLEX)) {
+                    if (subAttribConfig.hasChildren()) {
+                        String error = "A attribute of primitive type can not have sub attributes";
+                        throw new InternalErrorException(error);
+                    } else {
+                        buildSimpleAttributeSchema(subAttribConfig, attributeSchemas);
+                    }
+                } else {
+                    if (!(subAttribConfig.hasChildren())) {
+                        String error = "A attribute of complex type should have sub attributes";
+                        throw new InternalErrorException(error);
+                    } else {
+                        // need to build child schemas first
+                        buildComplexAttributeSchema(subAttribConfig, attributeSchemas, customConfig);
+                    }
+                }
+            }
+            // now all sub attributes must be already built
+            buildComplexSchema(config, attributeSchemas);
+        }
+    }
+
+    /*
+     * Has the logic to iterate through child attributes
```

Comment:
Should be a doc comment.

Reply:
fixed https://github.com/wso2/charon/pull/323/commits/6a773e3384a5db3baab94370f9f64a904834a48d